File: apps/reading/src/api/midnightOilUnattendedRecapCompose.ts
/**
 * Midnight Oil unattended recap compose (pure).
 *
 * Operator vision: set time + goals + price ceiling; swarm works unattended;
 * operator returns to a recap of progress, spend vs ceiling, and artifacts.
 *
 * This pure layer never re-launches workers.
 * live_execution_authorized always false.
 * store_mutated always false.
 */

export interface MidnightOilRecapGoal {
  goal_id: string;
  title: string;
  /** Caller-supplied status — never invented. */
  status: "pending" | "in_progress" | "done" | "blocked" | "skipped";
  /** Optional notes from run (caller-supplied). */
  notes?: string;
}

export interface MidnightOilUnattendedRecapInput {
  run_id: string;
  operator_id: string;
  work_minutes_planned: number;
  /** Actual minutes worked if known; null = unknown honesty. */
  work_minutes_actual: number | null;
  goals: MidnightOilRecapGoal[];
  /** Operator-approved ceiling for this run; null = unknown. */
  price_ceiling_usd: number | null;
  /** Settled or reported spend; null = unknown. */
  spend_usd: number | null;
  /** Caller-supplied artifact ids/paths only. */
  artifact_ids?: string[] | null;
  operator_ack: boolean;
}

export interface MidnightOilUnattendedRecapCompose {
  run_id: string;
  operator_id: string;
  goal_count: number;
  goals_done: number;
  goals_blocked: number;
  goals_pending: number;
  /** True when spend known and ceiling known and spend <= ceiling. */
  within_ceiling: boolean | null;
  /** True when ≥1 goal done or artifact present and operator_ack. */
  recap_ready: boolean;
  artifact_count: number;
  /** Always false — recap never re-authorizes live MO workers. */
  live_execution_authorized: false;
  /** Always false — pure recap does not persist. */
  store_mutated: false;
  notes: string[];
  authority: "midnight_oil_unattended_recap_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function requireNonNegFinite(value: unknown, name: string): number {
  if (typeof value !== "number" || !Number.isFinite(value) || value < 0) {
    throw new Error(`${name} must be a non-negative finite number`);
  }
  return value;
}

const VALID_STATUS = new Set([
  "pending",
  "in_progress",
  "done",
  "blocked",
  "skipped",
]);

/**
 * Compose an unattended MO run recap for the operator.
 * Never live-executes; never invents spend or goal outcomes.
 */
export function composeMidnightOilUnattendedRecap(
  input: MidnightOilUnattendedRecapInput,
): MidnightOilUnattendedRecapCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const run_id = requireNonEmpty(input.run_id, "run_id");
  const operator_id = requireNonEmpty(input.operator_id, "operator_id");
  const work_minutes_planned = requireNonNegFinite(
    input.work_minutes_planned,
    "work_minutes_planned",
  );
  if (work_minutes_planned <= 0) {
    throw new Error("work_minutes_planned must be > 0");
  }
  if (
    input.work_minutes_actual !== null &&
    input.work_minutes_actual !== undefined &&
    (typeof input.work_minutes_actual !== "number" ||
      !Number.isFinite(input.work_minutes_actual) ||
      input.work_minutes_actual < 0)
  ) {
    throw new Error(
      "work_minutes_actual must be non-negative finite number or null",
    );
  }
  const work_minutes_actual =
    input.work_minutes_actual === undefined ? null : input.work_minutes_actual;

  if (!Array.isArray(input.goals) || input.goals.length === 0) {
    throw new Error("goals must be a non-empty array");
  }

  const notes: string[] = [
    "live_execution_authorized=false — recap never re-launches MO workers",
    "store_mutated=false — recap is advisory snapshot only",
    "goal statuses and spend are caller-supplied only (no invent)",
  ];

  let goals_done = 0;
  let goals_blocked = 0;
  let goals_pending = 0;
  const seen = new Set<string>();

  for (let i = 0; i < input.goals.length; i++) {
    const g = input.goals[i];
    if (!g || typeof g !== "object") {
      throw new Error(`goals[${i}] must be an object`);
    }
    const id = requireNonEmpty(g.goal_id, `goals[${i}].goal_id`);
    if (seen.has(id)) {
      throw new Error(`duplicate goal_id: ${id}`);
    }
    seen.add(id);
    requireNonEmpty(g.title, `goals[${i}].title`);
    if (!VALID_STATUS.has(g.status)) {
      throw new Error(
        `goals[${i}].status must be pending|in_progress|done|blocked|skipped`,
      );
    }
    if (g.notes != null) {
      requireNonEmpty(g.notes, `goals[${i}].notes`);
    }
    if (g.status === "done") goals_done += 1;
    else if (g.status === "blocked") goals_blocked += 1;
    else if (g.status === "pending" || g.status === "in_progress") {
      goals_pending += 1;
    }
  }

  const goal_count = input.goals.length;
  notes.push(
    `goals=${goal_count} · done=${goals_done} · blocked=${goals_blocked} · pendingish=${goals_pending}`,
  );
  notes.push(`work_minutes_planned=${work_minutes_planned}`);
  if (work_minutes_actual === null) {
    notes.push("work_minutes_actual=null (unknown honesty)");
  } else {
    notes.push(`work_minutes_actual=${work_minutes_actual}`);
  }

  let within_ceiling: boolean | null = null;
  if (
    input.price_ceiling_usd !== null &&
    input.price_ceiling_usd !== undefined &&
    input.spend_usd !== null &&
    input.spend_usd !== undefined
  ) {
    if (
      typeof input.price_ceiling_usd !== "number" ||
      !Number.isFinite(input.price_ceiling_usd) ||
      input.price_ceiling_usd < 0
    ) {
      throw new Error("price_ceiling_usd must be non-negative finite when set");
    }
    if (
      typeof input.spend_usd !== "number" ||
      !Number.isFinite(input.spend_usd) ||
      input.spend_usd < 0
    ) {
      throw new Error("spend_usd must be non-negative finite when set");
    }
    within_ceiling = input.spend_usd <= input.price_ceiling_usd;
    notes.push(
      within_ceiling
        ? `within_ceiling=true · spend=${input.spend_usd} ≤ ceiling=${input.price_ceiling_usd}`
        : `within_ceiling=false · spend=${input.spend_usd} > ceiling=${input.price_ceiling_usd}`,
    );
  } else {
    notes.push(
      "within_ceiling=null — spend and/or ceiling unknown (no invent $0)",
    );
  }

  let artifact_count = 0;
  if (input.artifact_ids != null) {
    if (!Array.isArray(input.artifact_ids)) {
      throw new Error("artifact_ids must be an array when set");
    }
    const aseen = new Set<string>();
    for (let i = 0; i < input.artifact_ids.length; i++) {
      const aid = requireNonEmpty(
        input.artifact_ids[i],
        `artifact_ids[${i}]`,
      );
      if (aseen.has(aid)) {
        throw new Error(`duplicate artifact_id: ${aid}`);
      }
      aseen.add(aid);
    }
    artifact_count = aseen.size;
    notes.push(`artifact_count=${artifact_count}`);
  }

  const hasProgress = goals_done >= 1 || artifact_count >= 1;
  const recap_ready = input.operator_ack && hasProgress;
  if (!input.operator_ack) {
    notes.push("recap_ready=false — operator_ack required");
  } else if (!hasProgress) {
    notes.push(
      "recap_ready=false — no done goals or artifacts (no invent progress)",
    );
  } else {
    notes.push("recap_ready=true — operator may review unattended outcomes");
  }

  notes.push("live_execution_authorized=false");
  notes.push("store_mutated=false");

  return {
    run_id,
    operator_id,
    goal_count,
    goals_done,
    goals_blocked,
    goals_pending,
    within_ceiling,
    recap_ready,
    artifact_count,
    live_execution_authorized: false,
    store_mutated: false,
    notes,
    authority: "midnight_oil_unattended_recap_compose_advisory",
  };
}

export function formatMidnightOilUnattendedRecapSummary(
  c: MidnightOilUnattendedRecapCompose,
): string {
  return (
    `recap_ready=${c.recap_ready} · done=${c.goals_done}/${c.goal_count} · ` +
    `within_ceiling=${c.within_ceiling} · artifacts=${c.artifact_count} · ` +
    `live_execution_authorized=false`
  );
}

File: substrate/midnight_oil_unattended_recap_compose.py
"""Midnight Oil unattended recap compose (pure).

live_execution_authorized and store_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

GoalStatus = Literal["pending", "in_progress", "done", "blocked", "skipped"]
VALID_STATUS = frozenset(
    ("pending", "in_progress", "done", "blocked", "skipped")
)


class MidnightOilUnattendedRecapComposeError(ValueError):
    """Fail-closed validation for MO unattended recap."""


@dataclass(frozen=True)
class MidnightOilUnattendedRecapCompose:
    run_id: str
    operator_id: str
    goal_count: int
    goals_done: int
    goals_blocked: int
    goals_pending: int
    within_ceiling: bool | None
    recap_ready: bool
    artifact_count: int
    live_execution_authorized: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "run_id": self.run_id,
            "operator_id": self.operator_id,
            "goal_count": self.goal_count,
            "goals_done": self.goals_done,
            "goals_blocked": self.goals_blocked,
            "goals_pending": self.goals_pending,
            "within_ceiling": self.within_ceiling,
            "recap_ready": self.recap_ready,
            "artifact_count": self.artifact_count,
            "live_execution_authorized": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": "midnight_oil_unattended_recap_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise MidnightOilUnattendedRecapComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _require_nonneg_finite(value: object, *, field: str) -> float:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise MidnightOilUnattendedRecapComposeError(
            f"{field} must be a non-negative finite number"
        )
    f = float(value)
    if f != f or f < 0:
        raise MidnightOilUnattendedRecapComposeError(
            f"{field} must be a non-negative finite number"
        )
    return f


def compose_midnight_oil_unattended_recap(
    *,
    run_id: object,
    operator_id: object,
    work_minutes_planned: object,
    work_minutes_actual: object,
    goals: object,
    price_ceiling_usd: object,
    spend_usd: object,
    operator_ack: object,
    artifact_ids: object | None = None,
) -> MidnightOilUnattendedRecapCompose:
    """Compose MO unattended recap. Never re-launches workers."""
    if not isinstance(operator_ack, bool):
        raise MidnightOilUnattendedRecapComposeError(
            "operator_ack must be an explicit boolean"
        )
    rid = _require_nonempty(run_id, field="run_id")
    oid = _require_nonempty(operator_id, field="operator_id")
    planned = _require_nonneg_finite(
        work_minutes_planned, field="work_minutes_planned"
    )
    if planned <= 0:
        raise MidnightOilUnattendedRecapComposeError(
            "work_minutes_planned must be > 0"
        )
    if work_minutes_actual is not None:
        _require_nonneg_finite(
            work_minutes_actual, field="work_minutes_actual"
        )

    if not isinstance(goals, list) or len(goals) == 0:
        raise MidnightOilUnattendedRecapComposeError(
            "goals must be a non-empty array"
        )

    notes: list[str] = [
        "live_execution_authorized=false — recap never re-launches MO workers",
        "store_mutated=false — recap is advisory snapshot only",
        "goal statuses and spend are caller-supplied only (no invent)",
    ]

    goals_done = 0
    goals_blocked = 0
    goals_pending = 0
    seen: set[str] = set()
    for i, g in enumerate(goals):
        if not isinstance(g, dict):
            raise MidnightOilUnattendedRecapComposeError(
                f"goals[{i}] must be an object"
            )
        gid = _require_nonempty(g.get("goal_id"), field=f"goals[{i}].goal_id")
        if gid in seen:
            raise MidnightOilUnattendedRecapComposeError(
                f"duplicate goal_id: {gid}"
            )
        seen.add(gid)
        _require_nonempty(g.get("title"), field=f"goals[{i}].title")
        st = g.get("status")
        if st not in VALID_STATUS:
            raise MidnightOilUnattendedRecapComposeError(
                f"goals[{i}].status must be pending|in_progress|done|blocked|skipped"
            )
        if g.get("notes") is not None:
            _require_nonempty(g.get("notes"), field=f"goals[{i}].notes")
        if st == "done":
            goals_done += 1
        elif st == "blocked":
            goals_blocked += 1
        elif st in ("pending", "in_progress"):
            goals_pending += 1

    goal_count = len(goals)
    notes.append(
        f"goals={goal_count} · done={goals_done} · blocked={goals_blocked} · "
        f"pendingish={goals_pending}"
    )
    notes.append(f"work_minutes_planned={planned}")
    if work_minutes_actual is None:
        notes.append("work_minutes_actual=null (unknown honesty)")
    else:
        notes.append(f"work_minutes_actual={work_minutes_actual}")

    within_ceiling: bool | None = None
    if price_ceiling_usd is not None and spend_usd is not None:
        ceiling = _require_nonneg_finite(
            price_ceiling_usd, field="price_ceiling_usd"
        )
        spend = _require_nonneg_finite(spend_usd, field="spend_usd")
        within_ceiling = spend <= ceiling
        notes.append(
            f"within_ceiling=true · spend={spend} ≤ ceiling={ceiling}"
            if within_ceiling
            else f"within_ceiling=false · spend={spend} > ceiling={ceiling}"
        )
    else:
        notes.append(
            "within_ceiling=null — spend and/or ceiling unknown (no invent $0)"
        )

    artifact_count = 0
    if artifact_ids is not None:
        if not isinstance(artifact_ids, list):
            raise MidnightOilUnattendedRecapComposeError(
                "artifact_ids must be an array when set"
            )
        aseen: set[str] = set()
        for i, aid in enumerate(artifact_ids):
            a = _require_nonempty(aid, field=f"artifact_ids[{i}]")
            if a in aseen:
                raise MidnightOilUnattendedRecapComposeError(
                    f"duplicate artifact_id: {a}"
                )
            aseen.add(a)
        artifact_count = len(aseen)
        notes.append(f"artifact_count={artifact_count}")

    has_progress = goals_done >= 1 or artifact_count >= 1
    recap_ready = operator_ack and has_progress
    if not operator_ack:
        notes.append("recap_ready=false — operator_ack required")
    elif not has_progress:
        notes.append(
            "recap_ready=false — no done goals or artifacts (no invent progress)"
        )
    else:
        notes.append(
            "recap_ready=true — operator may review unattended outcomes"
        )

    notes.extend(
        ("live_execution_authorized=false", "store_mutated=false")
    )

    return MidnightOilUnattendedRecapCompose(
        run_id=rid,
        operator_id=oid,
        goal_count=goal_count,
        goals_done=goals_done,
        goals_blocked=goals_blocked,
        goals_pending=goals_pending,
        within_ceiling=within_ceiling,
        recap_ready=recap_ready,
        artifact_count=artifact_count,
        live_execution_authorized=False,
        store_mutated=False,
        notes=tuple(notes),
        authority="midnight_oil_unattended_recap_compose_advisory",
    )


__all__ = [
    "MidnightOilUnattendedRecapCompose",
    "MidnightOilUnattendedRecapComposeError",
    "compose_midnight_oil_unattended_recap",
]
